src/credentials/tests/TestChipCert.cpp

```diff
@@ -46,8 +47,24 @@ using namespace chip::Crypto;
 enum
 {
     kStandardCertsCount = 3,
-    kTestCertBufSize    = 2048, // Size of buffer needed to hold any of the test certificates
-                                // (in either CHIP or DER form), or to decode the certificates.
+
+    /**
+     * Maximum buffer size needed to hold any of the test certificates in CHIP TLV encoded form.
+     */
+    kTestCHIPCertBufSize = kOperationalCertificateMaxSize,
+
+    /**
+     * Maximum buffer size needed to hold any of the test certificates in DER encoded form.
+     * The same buffer size is also used when the certificate is decoded.
+     *
+     * TODO: Note that spec limits CHIP certificate size in DER form to 600 bytes.
+     *       Although some of the test certificates exceed 600 byte limit in DER form
+     *       their TLV encoded representation is still below the 400 byte limit.
+     *       For example, the largest Node02_05 certificate is 633 byte in DER form
+     *       and 355 bytes when encoded in TLV form.
+     *       Is the DER cert size limit important???
```

Comment:
Yes, in that if the spec has requirements that can't actually be satisfied in practice we have a problem....

@tcarmelveilleux

Reply:
I'm pretty sure the DER cert size is important because nodes hold certs in memory with this representation.

Reply:
@bzbarsky-apple  We can satisfy this requirement and I'll update those test certificates that currently exceed DER size limit of 600 bytes.

@msandstedt  I am not fully agree with your comment. Spec doesn't specify how certificate is stored in the device memory. I don't see any reason why devices should store certificate in the DER form. The only place where certificate should be use in DER form is when they generated by CAs because of the existing tooling and infrastructure. Once generated they should be converted to CHIP TLV form and be used and stored in CHIP TLV form, which is more efficient and compact.

Reply:
> @msandstedt I am not fully agree with your comment. Spec doesn't specify how certificate is stored in the device memory. I don't see any reason why devices should store certificate in the DER form. The only place where certificate should be use in DER form is when they generated by CAs because of the existing tooling and infrastructure. Once generated they should be converted to CHIP TLV form and be used and stored in CHIP TLV form, which is more efficient and compact.

This isn't practical.  We need to leverage libraries to use our certificates at runtime.  The most compact representation these libraries are likely to support is DER.

Reply:
Currently SDK implements CASE session establishment, which expects these certificates to be in CHIP TLV encoded form. What other usecases/libraries do you have in mind that will use these certificate in DER form?

Reply:
DER encoding is widely used in the SDK.  For instance:

[src/crypto/CHIPCryptoPAL.h](https://github.com/project-chip/connectedhomeip/blob/ed677f04ae8187098130f3701b5583f8db56a55e/src/crypto/CHIPCryptoPAL.h)
[src/controller/OperationalCredentialsDelegate.h](https://github.com/project-chip/connectedhomeip/blob/ed677f04ae8187098130f3701b5583f8db56a55e/src/controller/OperationalCredentialsDelegate.h)

Reply:
The spec is clear that OpCerts as used on the wire everywhere are TLV. The SDK impl shouldn't try to store DER on device, since it is much larger. However, the commissioners can use DER as they see fit, and possibly convert at runtime, but we have to be careful that initiating a CASE connection doesn't require DER encoding.

One of the largest flash contributions of crypto libraries on device is ASN.1 DER support for X.509  certs, when in reality, only a very small subsets of X.509 representations is allowed for operational certs.

The limit of 600 bytes for DER certs targeted the DAC and PAI cert, which have to be stored in device non-volatile storage, but never *processed*. There is no usecases for end-devices to ever actully parse the DAC, since it just has to be repeated to the commissioner, and the usage of the attestation key pair is actually just over some payloads where only the private key is necessary to be used, which is not even part of data stored as DER anywhere (in theory, since it's an implementation detail of P256Keypair).

Reply:
Also, re:

> I'm pretty sure the DER cert size is important because nodes hold certs in memory with this representation.

This is not 100% accurate. We should not have to force anyone to store DER certs in RAM on device. If an implementation wants to do this, they can. However, the SDK is unfortunately structured right now in such a way that there isn't much choice to using DER, even though the design of the spec, and incoming Weave technology where the TLV certs are use extensively, is to have native support of TLV certs being possible. In order to verify a TLV cert signature, you have to "pseudo" built it in RAM as DER since that's what the TBS data was originally. However, this is transient, not permanent. There is no need to permanently hold DER certs in RAM.